### PR TITLE
Fix: Android reported a huge 17 min delay for receiving paused events

### PR DESCRIPTION
**Pull request by matszczygiel**
This PR abstracts away the socket into a custom type that imposes timeouts automatically

### CODE REVIEW

File: drop-transfer/src/ws/client/socket.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    recv_last: Instant,
+    recv_timeout: Duration,
+    send_timeout: Duration,
+}
+
+impl WebSocket {
+    pub fn new(stream: WsStream, recv_timeout: Duration, send_timeout: Duration) -> Self {
+        Self {
+            stream,
+            recv_last: Instant::now(),
```
Comment (LukasPukenis) [inline drop-transfer/src/ws/client/socket.rs:23]: To be pedantic I think `Option<Instant>` would be slightly more correct

File: drop-transfer/src/ws/client/socket.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            recv_timeout,
+            send_timeout,
+        }
+    }
+
+    pub async fn send(&mut self, msg: Message) -> crate::Result<()> {
+        tokio::time::timeout(self.send_timeout, self.stream.send(msg))
+            .await
+            .map_err(|err| io::Error::new(io::ErrorKind::TimedOut, err))??;
```
Comment (LukasPukenis) [inline drop-transfer/src/ws/client/socket.rs:32]: I am curious why do we need 2x`?`? Can `map_err` fail to convert the message..?

File: drop-transfer/src/ws/client/socket.rs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    pub async fn recv(&mut self) -> crate::Result<Message> {
+        let timeout = self.recv_timeout.saturating_sub(self.recv_last.elapsed());
+
+        let msg = tokio::time::timeout(timeout, self.stream.next())
+            .await
+            .map_err(|err| io::Error::new(io::ErrorKind::TimedOut, err))?
+            .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "Emptry socket stream"))??;
+
+        Ok(msg)
```
Comment (LukasPukenis) [inline drop-transfer/src/ws/client/socket.rs:45]: Similar question to above, if the `await` resulted in a timeout error we map it to `TimedOut` but why do we need another `ok_or_else` for?

**Review by LukasPukenis — COMMENTED** [review]
Curious why those 17minute delays happened and how the existing timeout helps?

File: drop-transfer/src/ws/client/socket.rs (hunk 1 above)
Comment (matszczygiel) [reply drop-transfer/src/ws/client/socket.rs:23]: Oh, yeah. True. You gave me a thought that I'm not setting this field in the receive() method. My mistake

File: drop-transfer/src/ws/client/socket.rs (hunk 2 above)
Comment (matszczygiel) [reply drop-transfer/src/ws/client/socket.rs:32]: No, there are two levels of error here

Comment (LukasPukenis) [reply drop-transfer/src/ws/client/socket.rs:32]: Oh okay `stream.send` might error as well

File: drop-transfer/src/ws/client/socket.rs (hunk 1 above)
Comment (matszczygiel) [reply drop-transfer/src/ws/client/socket.rs]: Fixed

**Comment by matszczygiel** [comment]
@LukasPukenis the existing timeout was imposed on the `recv()` call. Yet Android managed to hang on the `send()` call

**Review by LukasPukenis — APPROVED** [review]

**Pull request merged**